Document how to implement file upload (#610)
front-commerce/developers.front-commerce.com: docs/advanced/server/file-upload.mdx

---
sidebar_position: 5

> title: File upload

Reviewer: ![2023-03-13_11-17](https://user-images.githubusercontent.com/75968/224673103-35afdb12-7ffb-48ee-8c59-54d2aab37d80.png)

```suggestion
title: File uploads
```

description:

>   This page explains how the file upload can be implemented in a Front-Commerce project.

Reviewer: ```suggestion
  This page explains how file uploads can be implemented in a Front-Commerce project.
```

---

import ContactLink from "@site/src/components/ContactLink";

To upload files to Front-Commerce, two main strategies are possible:

1. [Using an express based route to handle the upload](#with-an-express-based-route)
1. [Sending the file with a GraphQL mutation](#with-a-graphql-mutation)

Depending on the customer workflow and various technical details one or the
other can be more suited.

## With an express based route

> In that strategy, on the frontend you would typically have a form with [a file

Reviewer 2: ```suggestion
In this strategy, on the frontend you would typically have a form with [a file
```

input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file) so
that the user can pick a file. This form must have [the `enctype` attribute set
to
`multipart/form-data`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#enctype)
and must be configured to post to an express based route. For instance:

```jsx
<form action="/my/express/route" encType="multipart/form-data">
  <p>
    <label htmlFor="file">Choose a file</label>
    <input type="file" name="file" />
  </p>
  <p><button>Upload</button></p>
</form>
```

In this example, when the customer picks a file and submits the form, the file
will be sent to the URL `/my/express/route`. To create such route, please refer
to [the add custom endpoints to your server

> guide](/docs/advanced/server/add-http-endpoint). To ease the upload file

Reviewer 3: **nit:**
```suggestion
guide](/docs/advanced/server/add-http-endpoint). To ease the file upload```

Reviewer 3: Or 
```suggestion
guide](/docs/advanced/server/add-http-endpoint). To ease the uploaded file
```
If I misunderstood the meaning of this sentence.

handling, you can use [the express-fileupload
package](https://www.npmjs.com/package/express-fileupload).

:::note

The request to upload the file can also be issued with `fetch` or `XMLHttpRequest`

:::

## With a GraphQL mutation

> In that strategy, the file will be sent by passing it as a parameter to a GraphQL

Reviewer 2: ```suggestion
In this strategy, the file will be sent by passing it as a parameter to a GraphQL
```

mutation as a base64 encoded string. Here again, you would typically have a

> form but here instead of directly submitting it, a submit event handler would

Reviewer 2: ```suggestion
form but instead of directly submitting it, a submit event handler would
```

```jsx
import React from "react";
import { useLazyQuery } from "react-apollo";
import SendFileMutation from "./SendFileMutation.gql";

const UploadForm = () => {
  const [sendFile] = useLazyQuery(SendFileMutation);
  const uploadFile = (e) => {
    const file = e.target.elements.namedItem("file").files.item(0);
    const reader = new FileReader();
    reader.onload = () => {
      sendFile({ variables: { file: reader.result } });
    };
    reader.readAsDataURL(file);
    e.preventDefault();
  };
  return (
    <form onSubmit={uploadFile}>
      <p>
        <label htmlFor="file">Choose a file</label>
        <input type="file" name="file" />
      </p>
      <p>
        <button>Upload</button>
      </p>
    </form>
  );
};

export default UploadForm;
```

```graphql title=SendFileMutation.gql
mutation SendFileMutation($file: String!) {
  sendFileMutation(file: $file) {
    success
    errorMessage
    # and any other field depending on the actual result of the mutation
  }
}
```

For that example to work, the mutation `SendFileMutation` must be defined and
implemented in [a
custom GraphQL module](/docs/essentials/extend-the-graphql-schema).

---

> Feel free to <ContactLink /> if you have questions or remark on that

Reviewer 2: ```suggestion
Feel free to <ContactLink /> if you have questions or remark on this
```
